src/main/java/net/mcreator/blockly/data/BlocklyLoader.java

```diff
@@ -26,41 +28,66 @@ public static void init() {
 		INSTANCE = new BlocklyLoader();
 	}
 
-	private final ExternalBlockLoader procedureBlockLoader;
-	private final ExternalBlockLoader jsonTriggerLoader;
-	private final ExternalBlockLoader aitaskBlockLoader;
-	private final ExternalBlockLoader cmdargsBlockLoader;
-	private final ExternalBlockLoader featureBlockLoader;
+	private static final List<String> builtinCategories = new ArrayList<>() {{
+		add("other");
+		add("apis");
+		add("mcelements");
+		add("mcvariables");
+		add("customvariables");
+		add("logicloops");
+		add("logicoperations");
+		add("math");
+		add("text");
+		add("time");
+		add("advanced");
+		add("action");
+		add("aiadvanced");
+	}};
+	private final Map<String, ExternalBlockLoader> blockLoaders;
 	private final ExternalTriggerLoader externalTriggerLoader;
 
 	private BlocklyLoader() {
-		procedureBlockLoader = new ExternalBlockLoader("procedures");
-		aitaskBlockLoader = new ExternalBlockLoader("aitasks");
-		cmdargsBlockLoader = new ExternalBlockLoader("cmdargs");
+		blockLoaders = new LinkedHashMap<>();
+		addBlockLoader("procedures");
+		addBlockLoader("aitasks");
+		addBlockLoader("cmdargs");
+		addBlockLoader("jsontriggers");
+		addBlockLoader("features");
 		externalTriggerLoader = new ExternalTriggerLoader("triggers");
-		jsonTriggerLoader = new ExternalBlockLoader("jsontriggers");
-		featureBlockLoader = new ExternalBlockLoader("features");
 	}
 
-	public ExternalBlockLoader getProcedureBlockLoader() {
-		return procedureBlockLoader;
+	/**
+	 * Create a new {@link ExternalBlockLoader} to load blocks for a specific Blockly panel.
+	 *
+	 * @param name The folder's name where to load files
+	 */
+	public void addBlockLoader(String name) {
+		blockLoaders.put(name, new ExternalBlockLoader(name));
 	}
 
-	public ExternalBlockLoader getAITaskBlockLoader() {
-		return aitaskBlockLoader;
+	/**
+	 * Add a usable category for JSON file blocks that has been created inside the custom XML {@link ToolboxType} file.
+	 * All custom categories used by a {@link ToolboxType} have to be added before blocks are loaded.
+	 *
+	 * @param name The category's name as written inside the XML file (e.g. <i>&lt;custom-thisName/&gt;</i>).
+	 */
+	public static void addBuiltinCategory(String name) {
+		builtinCategories.add(name);
 	}
 
-	public ExternalBlockLoader getCmdArgsBlockLoader() {return cmdargsBlockLoader;}
+	public Map<String, ExternalBlockLoader> getAllBlockLoaders() {
+		return blockLoaders;
+	}
 
-	public ExternalTriggerLoader getExternalTriggerLoader() {
-		return externalTriggerLoader;
+	public static List<String> getBuiltinCategories() {
+		return builtinCategories;
 	}
 
-	public ExternalBlockLoader getJSONTriggerLoader() {
-		return jsonTriggerLoader;
+	public ExternalBlockLoader getBlockLoader(String name) {

```

Comment:
I think loading block loader should be done by BlocklyEditorType for key, not hardcoded string literal

Reply:
I will need to duplicate all methods because global triggers use the same system, but they can't use a BlocklyEditorType as they are not blockly blocks.

Reply:
Not sure what you mean, global triggers call `getBlockLoader` method?

Reply:
JSON triggers, sorry. However, the problem is still the same. They use the same system as the Java Blockly editors and so, using a `BlocklyEditorType` for methods will require to duplicate methods (to support json triggers) or to change this system, which will make the PR much bigger (or require a second PR).

Reply:
I see, however, I really don't like using hardcoded literals for this. I will think about it a bit and propose a solution, or alternatively we will use hardcoded strings for now

Reply:
You could add new BlocklyEditorType for JSON triggers, and make parameters that are not used null, which would indicate those features are not supported by this BlocklyEditorType